File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

bool solution(vector<string> phone_book) {
    // 접두어인 경우가 있으면 false
    // 없으면 true
    // 같은 전화번호 중복 없음
    
    // 1. 각 전화번호를 오름차순 정렬한다. (문자 사전 순으로 정렬됨)
    sort(phone_book.begin(), phone_book.end());
    // for(string p: phone_book) cout << p << endl;
    
    // 2. 각 원소가 자신의 바로 뒤 원소의 접두어가 될 수 있으면 false
    int length = phone_book.size();
    string cur, pre;
    for(int i=1; i<length; i++) {
        cur = phone_book[i];
        pre = phone_book[i-1];
        if (pre == cur.substr(0, pre.length())) return false;
    }
    
    return true;
}

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;

class Solution {
    
    static Map<String, Integer> pfs = new HashMap<>();
    
    public boolean solution(String[] phone_book) {
        // 해시를 쓴다면
        
        // 각 전화번호를 전화번호-인덱스 형태로 Map에 저장한다.
        int length = phone_book.length;
        for(int i=0; i<length; i++) {
            pfs.put(phone_book[i], i);
        }
        
        // 각 전화번호에 대해 다른 전화번호의 접두어 여부를 확인한다.
        String cur;
        for(int i=0; i<length; i++) {
            cur = phone_book[i];
            for(int j=0; j<cur.length(); j++) {
                if (pfs.containsKey(cur.substring(0, j)))
                    return false;
            }
        }
        
        return true;
    }
}